sql: add geospatial data types

otan opened the pull request:
This PR adds the minimum amount necessary for the new geospatial types
to work without breaking CI, with moderate testing of functionality.

None of the SRID handling or figure checks are in yet -- that will come
in the future.

Indexes are forbidden from creation, as we will only supported inverted
indexes for these geospatial types. Comparison operators depend on byte
comparisons of the EWKB.

Packages affected are as follows:
* geo/geopb: add new protobufs to handle figures and the SRID type. Also
modified Parse* to take in strings instead of geopb.WKT, as we should be
able to parse EWKBHexs/EWKTs using Parse. Added a few more utilities to
help with the build out.
* sql/types: add new handling structs that can deal with geometry and
geography base types.
* sql/types/oidext: new library that wraps the oid library which
contains the oids not found in lib/pq/oid, since they are "extensions".
We've arbitrarily set the OIDs for geospatial types to start at over
9000 at 90000.
* sql.y: added PARSING for GEOGRAPHY/GEOMETRY types. Needed to add POINT
and POLYGON as a col_name_keyword so the unimplemented syntax error for
POINT and POLYGON still remains. Note that GEOGRAPHY will automatically
set SRID to 4326, as per PostGIS behaviour.
* ccl/changefeedccl: change feeds of GEOGRAPHY/GEOMETRY types use the
EWKB representation.
* sql/sqlbase: encoding and decoding GEOGRAPHY/GEOMETRY types into disk.
* lint: added new inclusion for sql/types/oidext to use `T_` which is
form following.

Release note (sql change): Introduced the ability to parse, store and display
PostGIS style Geography and Geometry types.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47171)
<!-- Reviewable:end -->

otan commented:
+ @rohany for SQL side, + @sumeerbhola for geo side.

sumeerbhola reviewed (changes requested):
haven't read the non-geo code yet -- will look at it this evening.

> Reviewed 4 of 59 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-:-M4Px5y_C9fpZh14Lw7Y:bb7xujm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rohany, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4Pv0u9-gizWSjHFXra:-M4Pv0uAE2EeRU1SyI4u:b-7ubjk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/235c0e9e02608ae0d7a604f0c794267c65c3eeb3/pkg/geo/geo.go#L47)):*
> ```Go
> // makeSpatialObjectBase creates a spatialObjectBase from an unvalidated EWKB.
> func makeSpatialObjectBase(in geopb.UnvalidatedEWKB) (spatialObjectBase, error) {
> 	t, err := ewkb.Unmarshal(in)
> ```

does `Unmarshal` handle both endianness?
___
*[pkg/geo/geo.go, line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4Pw1uD6GYIY4W63RKh:-M4Pw1uD6GYIY4W63RKi:b-76r0lt) ([raw file](https://github.com/cockroachdb/cockroach/blob/235c0e9e02608ae0d7a604f0c794267c65c3eeb3/pkg/geo/geo.go#L118)):*
> ```Go
> 	if len(str) > 0 {
> 		// Parse as EWKB hex.
> 		if str[0] == '0' {
> ```

do you have links to where the hex and binary formats are documented? I don't understand why hex has to start with the '0' character and binary has to start with \00\01.

And is using `ewkbhex` and `ewkb` giving us some capability that is not part of `geos.WKTToWKB`? Can you elaborate on what the exact capabilities are?
___
*[pkg/geo/geopb/geopb.proto, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4PpC_q2K5tFDiDcqAb:-M4PpC_q2K5tFDiDcqAc:b-v5y67u) ([raw file](https://github.com/cockroachdb/cockroach/blob/47d88f0e49a31422cad4cca90157f040bd6fc14a/pkg/geo/geopb/geopb.proto#L15)):*
> ```protobuf
> option go_package = "geopb";
> 
> // Figure is the type of a spatial figure. Each of these corresponds to a
> ```

Is "figure" an generally accepted term? If not, I'd like to suggest "shape" as an alternative. The word "shape" is less used than "figure" in non-geospatial contexts, so is more specific. And in geospatial, shape is already used in the way we want, like "shape files". And PostGIS documentation (like https://postgis.net/docs/using_postgis_dbmanagement.html) sometimes uses "shape" (but I haven't seen "figure") e.g. "a type modifier that restricts the kind of shapes and dimensions allowed in the column".
___
*[pkg/geo/geopb/types.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4PtZRT7nKFW_1HmSQb:-M4PtZRT7nKFW_1HmSQc:bisa7k8) ([raw file](https://github.com/cockroachdb/cockroach/blob/47d88f0e49a31422cad4cca90157f040bd6fc14a/pkg/geo/geopb/types.go#L20)):*
> ```Go
> 	DefaultGeometrySRID = UnknownSRID
> 	// DefaultGeographySRID (aka 4326) is the GPS lat/lng we all know and love.
> 	DefaultGeographySRID = SRID(4326)
> ```

Can you add that 4326 (X, Y) coordinates are (lon, lat) expressed in degrees and with range [-180, 180] and [-90,90] respectively.
___
*[pkg/geo/geopb/types.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4PsndUCsHYWbMsABqQ:-M4PsndUCsHYWbMsABqR:b-i2yvdu) ([raw file](https://github.com/cockroachdb/cockroach/blob/47d88f0e49a31422cad4cca90157f040bd6fc14a/pkg/geo/geopb/types.go#L24)):*
> ```
> stored and represented as bare floats
> ```
how about: stored and represented using coordinates that are bare floats.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @sumeerbhola)

___
*[pkg/geo/geo.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4Pv0u9-gizWSjHFXra:-M4PxRGY9_XbiEc5Xhn6:b-371cqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/235c0e9e02608ae0d7a604f0c794267c65c3eeb3/pkg/geo/geo.go#L47)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

does `Unmarshal` handle both endianness?
</blockquote></details>

yes. the first byte in WKB indicates endianness.
___
*[pkg/geo/geo.go, line 118 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4Pw1uD6GYIY4W63RKh:-M4PxUsl6sXfuxtWrAVs:b-q351nx) ([raw file](https://github.com/cockroachdb/cockroach/blob/235c0e9e02608ae0d7a604f0c794267c65c3eeb3/pkg/geo/geo.go#L118)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

do you have links to where the hex and binary formats are documented? I don't understand why hex has to start with the '0' character and binary has to start with \00\01.

And is using `ewkbhex` and `ewkb` giving us some capability that is not part of `geos.WKTToWKB`? Can you elaborate on what the exact capabilities are?
</blockquote></details>

this is the equivalent of postgres's direct `'insert text here'::geometry`. 
see https://github.com/postgis/postgis/blob/master/postgis/lwgeom_inout.c#L85
it can also do geojson, but leave that for later. added a comment.
___
*[pkg/geo/geopb/geopb.proto, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4PpC_q2K5tFDiDcqAb:-M4Py6Vx9yhsGDn9bfPl:bevh2t5) ([raw file](https://github.com/cockroachdb/cockroach/blob/47d88f0e49a31422cad4cca90157f040bd6fc14a/pkg/geo/geopb/geopb.proto#L15)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is "figure" an generally accepted term? If not, I'd like to suggest "shape" as an alternative. The word "shape" is less used than "figure" in non-geospatial contexts, so is more specific. And in geospatial, shape is already used in the way we want, like "shape files". And PostGIS documentation (like https://postgis.net/docs/using_postgis_dbmanagement.html) sometimes uses "shape" (but I haven't seen "figure") e.g. "a type modifier that restricts the kind of shapes and dimensions allowed in the column".
</blockquote></details>

me and Dan decided on figure but happy to rename to shapes. Done.
___
*[pkg/geo/geopb/types.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4PtZRT7nKFW_1HmSQb:-M4Q--Kq0MWx0VeD-Q1U:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/47d88f0e49a31422cad4cca90157f040bd6fc14a/pkg/geo/geopb/types.go#L20)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add that 4326 (X, Y) coordinates are (lon, lat) expressed in degrees and with range [-180, 180] and [-90,90] respectively.
</blockquote></details>

Done.
___
*[pkg/geo/geopb/types.go, line 24 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4PsndUCsHYWbMsABqQ:-M4PyT3CCooi7YAN2RhR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/47d88f0e49a31422cad4cca90157f040bd6fc14a/pkg/geo/geopb/types.go#L24)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> ```
> stored and represented as bare floats
> ```
how about: stored and represented using coordinates that are bare floats.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/parser/sql.y:
```diff
@@ -7261,6 +7269,38 @@ simple_typename:
 | character_with_length
 | interval_type
 | postgres_oid
+| POINT error { return unimplementedWithIssueDetail(sqllex, 21286, "point") } // needed or else it generates a syntax error.
```
What kind of syntax are you trying to catch here?

otan replied:
because we added POINT to the keywords, it doesn't catch `CREATE TABLE TYPE (a POINT)` like we used to (somewhere deeper in sql layer), so moving that up here.

rohany commented:
You should add a version gate for creating objects of these types.

I don't have any particular comments -- all the added boilerplate looks correct, but I don't have any background on geospatial, so I don't know if the boilerplate itself is correct.

otan commented:
I'll be sure to add the version gate after we bump the version to 20.1!

sumeerbhola reviewed (changes requested):
> Reviewed 45 of 59 files at r1, 2 of 2 files at r2, 10 of 10 files at r3, 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan, @rohany, and @sumeerbhola)

___
*[pkg/geo/geo.go, line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QfDNV2yjDecIjTct6:-M4QfDNV2yjDecIjTct7:b29ik0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/geo/geo.go#L117)):*
> ```Go
> // See: https://github.com/postgis/postgis/blob/master/postgis/lwgeom_inout.c#L85.
> func parseAmbiguousTextToEWKB(str string) (geopb.EWKB, error) {
> 	// TODO(otan): check for SRID=x at the beginning.
> ```

Do you know of a spec for the WKB and EWKB encodings?
And is there a cheap way to convert one to the other. IIUC, we need WKB to pass back and forth to Geos functions. But our types are EWKB -- it would be convenient if we can just do slice concatenation to add the SRID prefix to WKB when receiving a WKB from Geos, and sub-slice to pass the WKB parameter to Geos.
___
*[pkg/geo/geo_test.go, line 177 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QiL4YDtHSY3NGLEM3:-M4QiL4YDtHSY3NGLEM4:bigo9iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/geo/geo_test.go#L177)):*
> ```Go
> 	}{
> 		{
> 			"0101000000000000000000F03F000000000000F03F",
> ```

is the test avoiding the SRID examples from https://github.com/postgis/postgis/blob/master/postgis/lwgeom_inout.c#L75-L81 because we don't handle SRID fully yet?
___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 22 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QlhrX8sRmYwQBJPzd:-M4QlhrX8sRmYwQBJPze:b-83a7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/logictest/testdata/logic_test/geospatial#L22)):*
> ```
> geom  GEOMETRY(POINT,4326)      true   NULL  ·  {}         false
> 
> statement error column bad_pk is of type geography and thus is not indexable
> ```

should "CREATE INDEX ON geo_table (geog)" and "CREATE INDEX ON geo_table (geom)" also fail?
___
*[pkg/sql/oidext/oidext.go, line 19 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4Qn9WE3zE8rRS9XFKt:-M4Qn9WE3zE8rRS9XFKu:b-waxna3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/oidext/oidext.go#L19)):*
> ```Go
> import "github.com/lib/pq/oid"
> 
> // OIDs in this block are extensions of postgres, thus having no official OID.
> ```

these are not official, but do they match the OIDs used by PostGIS?
___
*[pkg/sql/parser/sql.y, line 7275 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QnaBf5yRSilsVMa7h:-M4QnaBf5yRSilsVMa7i:b-pwrzvg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/parser/sql.y#L7275)):*
> ```Yacc
> | POLYGON error { return unimplementedWithIssueDetail(sqllex, 21286, "polygon") } // needed or else it generates a syntax error.
> 
> geo_figure:
> ```

nit: should this be geo_shape now?
___
*[pkg/sql/types/types.go, line 337 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QqQEuAic4JaOfnDa-:-M4QqQEuAic4JaOfnDa0:b-ok6xt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/types/types.go#L337)):*
> ```Go
> 			Locale: &emptyLocale,
> 			GeoMetadata: &GeoMetadata{
> 				SRID: geopb.DefaultGeometrySRID,
> ```

is the subtype type modifier being specified using GeoMetadata?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47171)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @sumeerbhola)

___
*[pkg/geo/geo.go, line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QfDNV2yjDecIjTct6:-M4RL6mV4HELcsdxU_Ge:bzbrd00) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/geo/geo.go#L117)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Do you know of a spec for the WKB and EWKB encodings?
And is there a cheap way to convert one to the other. IIUC, we need WKB to pass back and forth to Geos functions. But our types are EWKB -- it would be convenient if we can just do slice concatenation to add the SRID prefix to WKB when receiving a WKB from Geos, and sub-slice to pass the WKB parameter to Geos.

</blockquote></details>

OGC spec should have WKB.
EWKB -- wikipedia has the only difference in it -- https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry. It's not official but from what i've read, OGC respects it and won't touch that bitmask.
I think GEOS knows how to parse EWKB because they have an option of interpreting the SRID, I will test that in future. Otherwise, it's just removing 4 extra bytes and bitmasking another one.
___
*[pkg/geo/geo_test.go, line 177 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QiL4YDtHSY3NGLEM3:-M4RK_P-7HkEQm02iCN-:b-urkg45) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/geo/geo_test.go#L177)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is the test avoiding the SRID examples from https://github.com/postgis/postgis/blob/master/postgis/lwgeom_inout.c#L75-L81 because we don't handle SRID fully yet?
</blockquote></details>

yes. that's the big TODO on the parse function.
___
*[pkg/sql/logictest/testdata/logic_test/geospatial, line 22 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QlhrX8sRmYwQBJPzd:-M4RKd4GAZA6hlkpaJn3:b-84qikj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/logictest/testdata/logic_test/geospatial#L22)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

should "CREATE INDEX ON geo_table (geog)" and "CREATE INDEX ON geo_table (geom)" also fail?
</blockquote></details>

Yes. I've added tests for these cases.
___
*[pkg/sql/oidext/oidext.go, line 19 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4Qn9WE3zE8rRS9XFKt:-M4RKuo5BIZ7RIdKsrlf:b-933xu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/oidext/oidext.go#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

these are not official, but do they match the OIDs used by PostGIS?
</blockquote></details>

PostGIS dynamically loads OIDs into the next available OID in `pg_types`, since postgres supports dynamic types. The OID of geo on your database may be different to mine.
___
*[pkg/sql/parser/sql.y, line 7275 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QnaBf5yRSilsVMa7h:-M4RL3058Z9icEQTGW-c:b-pc8fji) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/parser/sql.y#L7275)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: should this be geo_shape now?
</blockquote></details>

yes! missed a spot, oops.
___
*[pkg/sql/types/types.go, line 337 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QqQEuAic4JaOfnDa-:-M4RLZNK347jJjshFkyC:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/sql/types/types.go#L337)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is the subtype type modifier being specified using GeoMetadata?
</blockquote></details>

Yes.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4QfDNV2yjDecIjTct6:-M4RM6eWDvPqzb8y1AX1:bmiegrw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a75c1587b79e54484746ef0665a77a97f286a894/pkg/geo/geo.go#L117)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

OGC spec should have WKB.
EWKB -- wikipedia has the only difference in it -- https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry. It's not official but from what i've read, OGC respects it and won't touch that bitmask.
I think GEOS knows how to parse EWKB because they have an option of interpreting the SRID, I will test that in future. Otherwise, it's just removing 4 extra bytes and bitmasking another one.
</blockquote></details>

oh wikipedia does not have the difference.
using their example -- the difference is that the 4 byte integer uses 0x2 to mean "SRID", and an 4 extra bytes after that 4 byte integer has the SRID.


<!-- Sent from Reviewable.io -->

jordanlewis commented on pkg/sql/parser/sql.y:
```diff
@@ -7261,6 +7269,38 @@ simple_typename:
+| MULTIPOLYGON { $$.val = geopb.Shape_MultiPolygon }
+| MULTILINESTRING { $$.val = geopb.Shape_MultiLineString }
+| MULTIPOINT { $$.val = geopb.Shape_MultiPoint }
+| GEOMETRY { $$.val = geopb.Shape_Geometry }
```
I'd expect there to be parse tests for every one of these cases, to prevent regressions

jordanlewis commented on pkg/sql/sem/tree/datum.go:
```diff
@@ -3891,6 +4061,12 @@ func NewDefaultDatum(evalCtx *EvalContext, t *types.T) (d Datum, err error) {
 		return dNullJSON, nil
 	case types.TimeTZFamily:
 		return dZeroTimeTZ, nil
+	case types.GeometryFamily, types.GeographyFamily:
```
I think you have to put a value for this - it doesn't matter what. This was added recently to work around an issue with column backfills, I think...

jordanlewis commented on pkg/sql/types/types.go:
```diff
@@ -1066,7 +1131,7 @@ func (t *T) Name() string {
 //
 func (t *T) PGName() string {
-	name, ok := oid.TypeName[t.Oid()]
+	name, ok := oidext.TypeName(t.Oid())
```
I don't have a better suggestion right now, but I'm nervous that people will forget to call the oidext version when they're supposed to... anything we can do about this? Maybe we just want to wrap the oid package with our own oid package and ban imports of the other oid package?

jordanlewis reviewed (commented):
Sorry for the drive by - I couldn't resist!

rohany commented on pkg/sql/logictest/testdata/logic_test/geospatial:
```diff
@@ -0,0 +1,32 @@
+statement ok
+CREATE TABLE geo_table(id int primary key, geog geography(geometry, 4326), geom geometry(point, 4326))
```
They would need to be in the not first column family to have the slightly different encoding

otan commented on pkg/sql/logictest/testdata/logic_test/geospatial:
```diff
@@ -0,0 +1,32 @@
+statement ok
+CREATE TABLE geo_table(id int primary key, geog geography(geometry, 4326), geom geometry(point, 4326))
```
Yes, I've omitted it for now because we haven't released 20.1 yet which makes testing that hard. https://github.com/orgs/cockroachdb/projects/6#card-35995350

otan committed:
sql: add geospatial data types

This PR adds the minimum amount necessary for the new geospatial types
to work without breaking CI, with moderate testing of functionality.

None of the SRID handling or figure checks are in yet -- that will come
in the future.

Indexes are forbidden from creation, as we will only supported inverted
indexes for these geospatial types. Comparison operators depend on byte
comparisons of the EWKB.

Packages affected are as follows:
* geo/geopb: add new protobufs to handle figures and the SRID type. Also
modified Parse* to take in strings instead of geopb.WKT, as we should be
able to parse EWKBHexs/EWKTs using Parse. Added a few more utilities to
help with the build out.
* sql/types: add new handling structs that can deal with geometry and
geography base types.
* sql/types/oidext: new library that wraps the oid library which
contains the oids not found in lib/pq/oid, since they are "extensions".
We've arbitrarily set the OIDs for geospatial types to start at over
9000 at 90000.
* sql.y: added PARSING for GEOGRAPHY/GEOMETRY types. Needed to add POINT
and POLYGON as a col_name_keyword so the unimplemented syntax error for
POINT and POLYGON still remains. Note that GEOGRAPHY will automatically
set SRID to 4326, as per PostGIS behaviour.
* ccl/changefeedccl: change feeds of GEOGRAPHY/GEOMETRY types use the
EWKB representation.
* sql/sqlbase: encoding and decoding GEOGRAPHY/GEOMETRY types into disk.
* lint: added new inclusion for sql/types/oidext to use `T_` which is
form following.

Release note (sql change): Introduced the ability to parse, store and display
PostGIS style Geography and Geometry types.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 59 files at r1, 2 of 3 files at r5, 7 of 7 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-:-M4UwKlt1GvE4xTq56q_:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @otan, @rohany, and @sumeerbhola)

___
*[pkg/sql/sem/tree/datum.go, line 4064 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4MYVVcWUfys_rFBCHr-r5-4064:-M4UvUoD1wErm0u57qEr:b-r5xtlh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae42e1157c38a9f78054e286feb4722674e7b3/pkg/sql/sem/tree/datum.go#L4064)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

hm, in that case, i'd make this only NULLABLE.
I'm going to make this a TODO since this PR is too big as is and I'd like to get this in before doing that.
</blockquote></details>

I'm not sure I understood this, but it sounds like we are going to force geospatial columns to be NULLABLE so that schema changes that add/remove columns don't need to construct a default value when the user did not provide one -- that seems harsh.


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/sem/tree/datum.go, line 4064 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/47171#-M4MYVVcWUfys_rFBCHr-r5-4064:-M4Uyzjj7JsI_Y-eVzwi:b9fu3y5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae42e1157c38a9f78054e286feb4722674e7b3/pkg/sql/sem/tree/datum.go#L4064)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm not sure I understood this, but it sounds like we are going to force geospatial columns to be NULLABLE so that schema changes that add/remove columns don't need to construct a default value when the user did not provide one -- that seems harsh.
</blockquote></details>

that's the same thing PostGIS does. making a column NOT NULL and adding to it is strictly forbidden, because they add an extra check constraint in there if you decide to do that:

```
otan=# create temp table hello2 (a int, b geography NOT NULL);
CREATE TABLE
otan=# insert into hello2 (a) values (1);
ERROR:  null value in column "b" violates not-null constraint
DETAIL:  Failing row contains (1, null).
otan=# \d hello2
              Table "pg_temp_3.hello2"
 Column |   Type    | Collation | Nullable | Default
--------+-----------+-----------+----------+---------
 a      | integer   |           |          |
 b      | geography |           | not null |

```


<!-- Sent from Reviewable.io -->

otan commented:
TFTRs+

bors=sumeerbhola

otan commented:
oops

bors r=sumeerbhola
